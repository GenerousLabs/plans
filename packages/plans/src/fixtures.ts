import mockFs from 'mock-fs';

export const bob = {
  slug: 'bob',
  data: { name: 'Bob' },
  frontmatter: `---
name: Bob
---`,
  markdown: '',
};
export const charlie = {
  slug: 'charlie',
  data: { name: 'Charlie' },
  frontmatter: `---
name: Charlie
---`,
  markdown: '',
};
export const aliceMessage = {
  data: { sender: 'alice', dateTimestampSeconds: 1590595620 },
  frontmatter: `---
sender: alice
dateTimestampSeconds: 1590595620
---`,
  markdown: 'Are you up for sharing this with me?',
};
export const charlieMessage = {
  data: { sender: 'charlie', dateTimestampSeconds: 1590595860 },
  frontmatter: `---
sender: charlie
dateTimestampSeconds: 1590595860
---`,
  markdown: "Would love to share this with you if you're interested.",
};

export const spotify = {
  slug: 'spotify',
  data: { name: 'Spotify' },
  frontmatter: `---
name: Spotify
---`,
  markdown: `I have two spots on my family plan.
`,
};
export const nordvpn = {
  slug: 'nordvpn',
  data: { name: 'NordVPN' },
  frontmatter: `---
name: NordVPN
---`,
  markdown: `I can use 6 devices at the same time
`,
};
export const omgyes = {
  slug: 'omgyes',
  data: { name: 'OMG Yes' },
  frontmatter: `---
name: OMG Yes
---`,
  markdown: `OMG Yes subscription
`,
};

export const joinFrontmatter = ({
  frontmatter,
  markdown,
}: {
  frontmatter: string;
  markdown: string;
}) => {
  return `${frontmatter}\n${markdown}`;
};

export const mockFilesystem = () => {
  mockFs({
    'alice/': {
      [bob.slug]: {
        'index.md': joinFrontmatter(bob),
        plans: {
          [spotify.slug]: {
            'index.md': joinFrontmatter(spotify),
          },
          [nordvpn.slug]: {
            'index.md': joinFrontmatter(nordvpn),
          },
        },
      },
      charlie: {
        'index.md': joinFrontmatter(charlie),
        plans: {
          [omgyes.slug]: {
            'index.md': joinFrontmatter(omgyes),
            [`message-${aliceMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              aliceMessage
            ),
            [`message-${charlieMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              charlieMessage
            ),
          },
        },
      },
    },
  });
};

export const mockFilesystemRestore = () => {
  mockFs.restore();
};
