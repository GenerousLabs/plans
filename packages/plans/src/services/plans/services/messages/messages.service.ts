import matter from 'gray-matter';
import { FS } from '../../../../shared.types';

export type Message = {
  id: string;
  planId: string;
  slug: string;
  sender: string;
  contentMarkdown: string;
};

type MessageFrontmatter = {
  sender: string;
  dateTimestampSeconds: number;
};

const parseMessageDataFromFrontmatter = (input: {
  [key: string]: any;
}): MessageFrontmatter => {
  const { sender, dateTimestampSeconds } = input;
  if (typeof sender !== 'string') {
    throw new Error('Failed to find sender for message. #gtNSF1');
  }
  if (typeof dateTimestampSeconds !== 'number') {
    throw new Error('Failed to find dateTimestampSeconds for message. #qD6aB3');
  }
  return { sender, dateTimestampSeconds };
};

export const parseMessageFromMarkdown = async (input: string) => {
  const parsed = matter(input);
  const { content } = parsed;
  const data = parseMessageDataFromFrontmatter(parsed.data);
  return { contentMarkdown: content, ...data };
};

/**
 * Return the contents of a single message file.
 */
export const getMessageDataFromPath = async ({
  fs,
  path,
}: {
  fs: FS;
  path: string;
}) => {
  const markdown = await fs.promises.readFile(path, {
    encoding: 'utf8',
  });

  const data = await parseMessageFromMarkdown(markdown);

  return data;
};
