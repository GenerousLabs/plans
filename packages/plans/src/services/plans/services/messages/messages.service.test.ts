import fs from 'fs';
import {
  afterEach,
  beforeEach,
  describe,
  expect,
  it,
} from 'jest-without-globals';
import * as f from '../../../../fixtures';
import { getMessageDataFromPath } from './messages.service';

describe('messages.service', () => {
  beforeEach(() => {
    f.mockFilesystem();
  });

  afterEach(() => {
    f.mockFilesystemRestore();
  });

  describe('getMessageDataFromPath()', () => {
    it('Fetches message data correctly #OtOeRY', async () => {
      expect(
        await getMessageDataFromPath({
          fs,
          path: 'alice/charlie/plans/charlie/omgyes/message-1590595620.md',
        })
      ).toMatchSnapshot();
    });

    it('Throws on empty file', async () => {
      await expect(
        getMessageDataFromPath({
          fs,
          path: 'elena/index.md',
        })
      ).rejects.toThrow();
    });
  });
});
