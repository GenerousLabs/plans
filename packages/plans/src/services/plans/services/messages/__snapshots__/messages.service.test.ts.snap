// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messages.service getMessageDataFromPath() Fetches message data correctly #OtOeRY 1`] = `
Object {
  "contentMarkdown": "Are you up for sharing this with me?",
  "dateTimestampSeconds": 1590595620,
  "sender": "alice",
}
`;
